slides/k8s/staticpods.md

```diff
@@ -0,0 +1,239 @@
+# Static pods
+
+- Pods are usually created indirectly, through another resource:
+
+  Deployment, Daemon Set, Job, Stateful Set ...
+
+- They can also be created directly
+
+- This can be done by writing YAML and using `kubectl apply` or `kubectl create`
+
+- Some resources (not all of them) can be created with `kubectl run`
+
+- Creating a resource with `kubectl` requires the API to be up
+
+- If we want to run the API server (and its dependencies) on Kubernetes itself ...
+
+  ... how can we create API pods (and other resources) when the API is not up yet?
+
+---
+
+## In theory
+
+- Each component of the control plane can be replicated
+
+- We could set up the control plane outside of the cluster
+
+- Then, once the cluster is up, create replicas running on the cluster
+
+- Finally, remove the replicas that are running outside of the cluster
+
+*What could possibly go wrong?*
+
+---
+
+## Sawing off the branch you're sitting on
+
+- What if anything goes wrong?
+
+  (During the setup or at a later point)
+
+- Worst case scenario, we might need to:
+
+  - set up a new control plane (outside of the cluster)
+  
+  - restore a backup from the old control plane
+  
+  - move the new control plane to the cluster (again)
+
+- This doesn't sound like a great experience
+
+---
+
+## Static pods to the rescue
+
+- Pods are started by kubelet (an agent running on every node)
+
+- To know which pods it should run, the kubelet queries the API server
+
+- The kubelet can also get a list of *static pods* from:
+
+  - a directory containing one (or multiple) *manifests*, and/or
+  
+  - a URL (serving a *manifest*)
+
+- These "manifests" are basically YAML definitions
+
+  (As produced by `kubectl get pod my-little-pod -o yaml --export`)
+
+---
+
+## Static pods are dynamic
+
+- Kubelet will periodically reload the manifests
+
+- It will start/stop pods accordingly
+
+  (i.e. it is not necessary to restart the kubelet after updating the manifests)
+
+- When connected to the Kubernetes API, the kubelet will create *mirror pods*
+
+- Mirror pods are copies of the static pods
+
+  (so they can be seen with e.g. `kubectl get pods`)
+
+---
+
+## Bootstrapping a cluster with static pods
+
+- We can run control plane components with these static pods
+
+- They don't need the API to be up (just the kubelet)
```

Comment:
"They don't need the API to be available"

Reply:
Or does this mean "They don't need the API to be available for them to be launched"?

Reply:
Good call; I'm suggesting 

- They can start without requiring access to the API server

- Once they are up and running, the API becomes available